packages/marshal/src/encodeToJSON.js

```diff
@@ -0,0 +1,367 @@
+// which may migrate here. The main external difference is that
+// `encodePassable` goes directly to string, whereas `encodeToJSON`
+// encodes to a raw JSON tree and leaves to the caller (`marshal.js`)
+// to stringify it.
+
+import { passStyleOf } from './passStyleOf.js';
+
+import { ErrorHelper } from './helpers/error.js';
+import { makeTagged } from './makeTagged.js';
+import { isObject, getTag } from './helpers/passStyle-helpers.js';
+import {
+  assertPassableSymbol,
+  nameForPassableSymbol,
+  passableSymbolForName,
+} from './helpers/symbol.js';
+
+/** @typedef {import('./types.js').MakeMarshalOptions} MakeMarshalOptions */
+/** @template Slot @typedef {import('./types.js').ConvertSlotToVal<Slot>} ConvertSlotToVal */
+/** @template Slot @typedef {import('./types.js').ConvertValToSlot<Slot>} ConvertValToSlot */
+/** @template Slot @typedef {import('./types.js').Serialize<Slot>} Serialize */
+/** @template Slot @typedef {import('./types.js').Unserialize<Slot>} Unserialize */
+/** @typedef {import('./types.js').Passable} Passable */
+/** @typedef {import('./types.js').InterfaceSpec} InterfaceSpec */
+/** @typedef {import('./types.js').Encoding} Encoding */
+/** @typedef {import('./types.js').Remotable} Remotable */
+
+const { ownKeys } = Reflect;
+const { isArray } = Array;
+const {
+  getOwnPropertyDescriptors,
+  defineProperties,
+  is,
+  fromEntries,
+  freeze,
+} = Object;
+const { details: X, quote: q } = assert;
+
+/**
+ * Special property name that indicates an encoding that needs special
+ * decoding.
+ */
+const QCLASS = '@qclass';
+export { QCLASS };
+
+/**
+ * @typedef {object} EncodeToJSONOptionsRecord
+ * @property {(remotable: object) => Encoding} encodeRemotableToJSON
+ * @property {(promise: object) => Encoding} encodePromiseToJSON
+ * @property {(error: object) => Encoding} encodeErrorToJSON
+ */
+
+/**
+ * @typedef {Partial<EncodeToJSONOptionsRecord>} EncodeToJSONOptions
+ */
+
+/**
+ * @param {EncodeToJSONOptions=} encodeOptions
+ * @returns {(passable: Passable) => Encoding}
+ */
+export const makeEncodeToJSON = ({
+  encodeRemotableToJSON = rem => assert.fail(X`remotable unexpected: ${rem}`),
+  encodePromiseToJSON = prom => assert.fail(X`promise unexpected: ${prom}`),
+  encodeErrorToJSON = err => assert.fail(X`error unexpected: ${err}`),
+} = {}) => {
+  /**
+   * Must encode `val` into plain JSON data *canonically*, such that
+   * `JSON.stringify(encode(v1)) === JSON.stringify(encode(v1))`
+   * For each copyRecord, we only accept string property names,
+   * not symbols. The encoded form the sort
+   * order of these names must be the same as their enumeration
+   * order, so a `JSON.stringify` of the encoded form agrees with
+   * a canonical-json stringify of the encoded form.
+   *
+   * @param {Passable} passable
+   * @returns {Encoding}
+   */
+  const encodeToJSON = passable => {
+    if (ErrorHelper.canBeValid(passable)) {
+      return encodeErrorToJSON(passable);
+    }
```

Comment:
No. I moved it outside the recursion with the following comment:

```js
      // We pull out this special case to accommodate errors that are not
      // valid Passables. For example, because they're not frozen.
      // The special case can only ever apply at the root, and therefore
      // outside the recursion, since an error could only be deeper in
      // a passable structure if it were passable.
      //
      // We pull out this special case because, for these errors, we're much
      // more interested in reporting whatever diagnostic information they
      // carry than we are about reporting problems encountered in reporting
      // this information.
```

Reply:
Resolving